### PR TITLE
[SPARK-42349][PYTHON] Support pandas cogroup with multiple df

### PR DESCRIPTION
**Pull request by santosh-d3vpl3x**
### What changes were proposed in this pull request?
Alternative approach to #39902. This PR removes conditional presence of first arg as grouping key and makes it mandatory. @EnricoMi suggested this could be one potential approach. This approach doubles down on "Explicit better than implicit" and that leads to reduction in cognitive overload.

All the remaining details remain more or less the same. Note: this is a breaking change to this experimental API.

Pandas cogroup UDF with applyInPandas currently support two dataframes. This is already very useful but limits us both API wise and efficiency wise when we have to use multiple DFs with cogroup.applyInPandas. The PR here is to support multiple DFs in cogroup.

```python
df1 = spark.createDataFrame(
    [(20000101, 1, 1.0), (20000101, 2, 2.0), (20000102, 1, 3.0), (20000102, 2, 4.0)],
    ("time", "id", "v1"))
df2 = spark.createDataFrame(
    [(20000101, 1, "x"), (20000101, 2, "y")],
    ("time", "id", "v2"))
df3 = spark.createDataFrame(
    [(20000101, 1, "asd"), (20000101, 2, "d")],
    ("time", "id", "v3"))
df4 = spark.createDataFrame(
    [(20000101, 1, "v"), (20000101, 2, "g")],
    ("time", "id", "v4"))
def asof_join(key, df1, df2, df3, df4):
    df12 = pd.merge_asof(df1, df2, on="time", by="id")
    df123 = pd.merge_asof(df12, df3, on="time", by="id")
    df1234 = pd.merge_asof(df123, df4, on="time", by="id")
    return df1234
df1.groupby("id").cogroup(df2.groupby("id"), df3.groupby("id"), df4.groupby("id")).applyInPandas(
    asof_join, schema="time int, id int, v1 double, v2 string, v3 string, v4 string"
).show()

+--------+---+---+---+---+---+
|    time| id| v1| v2| v3| v4|
+--------+---+---+---+---+---+
|20000101|  1|1.0|  x|asd|  v|
|20000102|  1|3.0|  x|asd|  v|
|20000101|  2|2.0|  y|  d|  g|
|20000102|  2|4.0|  y|  d|  g|
+--------+---+---+---+---+---+
```
image

### Why are the changes needed?
Currently pyspark support `cogroup.applyInPandas` with only 2 dataframes. The improvement request is to support multiple dataframes with variable arity. We have cases where we want to pass multiple(20 to 30) data frames to `cogroup.applyInPandas` function which then are used to either combine them together for further processing in spark or emit excel files or combined graphs. Workarounds are possible but they are very inefficient and error-prone.

### Does this PR introduce _any_ user-facing change?
The previous experimental implementation is awkward. The UDF can accept either 2 or 3 args. If 2 args are passed in the UDF then UDF receives 2 cogrouped DFs, if 3 args are passed then UDF receives 2 cogrouped DFs with grouping key as the first arg. he previous API is limiting and has implications on new proposed change. There is no clear way of distinguishing whether 3 args are for 3 DFs or 2 DFs + 1 grouping key. 
This change removes this behaviour. In this change, we always pass key as first arg to the UDF. 
Breaking change:
```python
def test_udf(df1, df2):
    return df1

df1.groupby("id").cogroup(df2.groupby("id")).applyInPandas(
    test_udf, "time int, id int, v1 double").show()

```
BEFORE:
```
+--------+---+---+
|    time| id| v1|
+--------+---+---+
|20000101|  1|1.0|
|20000102|  1|3.0|
|20000101|  2|2.0|
|20000102|  2|4.0|
+--------+---+---+
```
AFTER:
```
Traceback (most recent call last):
  File "previous.py", line 64, in <module>
    df1.groupby("id").cogroup(df2.groupby("id")).applyInPandas(
  File "/workspace/spark/python/lib/pyspark.zip/pyspark/sql/pandas/group_ops.py", line 459, in applyInPandas
  File "/workspace/spark/python/lib/pyspark.zip/pyspark/sql/pandas/functions.py", line 387, in pandas_udf
  File "/workspace/spark/python/lib/pyspark.zip/pyspark/sql/pandas/functions.py", line 452, in _create_pandas_udf
ValueError: Invalid function: the function in cogroup.applyInPandas must contain vararg or must take three or more argument with key being first
```
To make it work, change the UDF:
```python
def test_udf(key, df1, df2):
    return df1

df1.groupby("id").cogroup(df2.groupby("id")).applyInPandas(
    test_udf, "time int, id int, v1 double").show()
```

### CODE REVIEW

File: python/pyspark/sql/pandas/group_ops.py [hunk 1]
```diff
@@ -368,23 +368,23 @@ class PandasCogroupedOps:
     This API is experimental.
     """
 
-    def __init__(self, gd1: "GroupedData", gd2: "GroupedData"):
+    def __init__(self, gd1: "GroupedData", *gds: "GroupedData"):
         self._gd1 = gd1
-        self._gd2 = gd2
+        self._gds = gds
```
Comment (EnricoMi) [inline python/pyspark/sql/pandas/group_ops.py:373]: assert `gds` is long enough

File: python/pyspark/sql/pandas/group_ops.py [hunk 2]
```diff
@@ -397,12 +397,13 @@ def applyInPandas(
...
         func : function
-            a Python native function that takes two `pandas.DataFrame`\\s, and
-            outputs a `pandas.DataFrame`, or that takes one tuple (grouping keys) and two
+            a Python native function that takes one tuple (grouping keys) and two or more
             ``pandas.DataFrame``\\s, and outputs a ``pandas.DataFrame``.
         schema : :class:`pyspark.sql.types.DataType` or str
             the return type of the `func` in PySpark. The value can be either a
             :class:`pyspark.sql.types.DataType` object or a DDL-formatted type string.
+        pass_key : : bool
+             Used to pass key to the UDF when cogrouped over more than 2 dataframes.
```
Comment (EnricoMi) [inline python/pyspark/sql/pandas/group_ops.py:406]: leaked from the other PR, I think...
```suggestion
```

File: python/pyspark/sql/pandas/group_ops.py [hunk 3]
```diff
@@ -413,40 +414,35 @@ def applyInPandas(
...
         |    time| id| v1| v2|
         +--------+---+---+---+
         |20000101|  1|1.0|  x|
         |20000102|  1|3.0|  x|
         +--------+---+---+---+
-
+        >>> df3 = spark.createDataFrame(
+        ...     [(20000101, 1, "x"), (20000101, 2, "y")],
+        ...     ("time", "id", "v3"))
+        >>> def asof_join_multiple(_, *dfs):
```
Comment (EnricoMi) [inline python/pyspark/sql/pandas/group_ops.py:433]: example should mention `key` argument, even if it is not used inside
```suggestion
        >>> def asof_join_multiple(key, *dfs):
```

File: python/pyspark/sql/pandas/group_ops.py [hunk 4]
```diff
@@ -463,12 +459,20 @@ def applyInPandas(
...
+            returnType=schema,
+            functionType=PythonEvalType.SQL_COGROUPED_MAP_PANDAS_UDF,
         )  # type: ignore[call-overload]
 
-        all_cols = self._extract_cols(self._gd1) + self._extract_cols(self._gd2)
-        udf_column = udf(*all_cols)
-        jdf = self._gd1._jgd.flatMapCoGroupsInPandas(self._gd2._jgd, udf_column._jc.expr())
+        all_cols = self._extract_cols(self._gd1)
+        for gd in self._gds:
+            all_cols.extend(self._extract_cols(gd))
```
Comment (EnricoMi) [inline python/pyspark/sql/pandas/group_ops.py:469]: Could be simplified to
```suggestion
        all_cols = [self._extract_cols(gd) for gd in [self._gd1] + self._gds]
```

File: python/pyspark/sql/pandas/group_ops.py [hunk 5]
```diff
@@ -463,12 +459,20 @@ def applyInPandas(
...
+        udf_column_expr = udf(*all_cols)._jc.expr()
+        assert self._gd1.session.sparkContext._jvm is not None
+        all_jgds = self._gd1.session.sparkContext._jvm.PythonUtils.toSeq(
+            [gd._jgd for gd in self._gds]
+        )
+        jdf = self._gd1._jgd.flatMapCoGroupsInPandas(all_jgds, udf_column_expr)
```
Comment (EnricoMi) [inline python/pyspark/sql/pandas/group_ops.py:475]: `all` is misleading, as this refers only to `self._gds`:
```suggestion
        jgds = self._gd1.session.sparkContext._jvm.PythonUtils.toSeq(
            [gd._jgd for gd in self._gds]
        )
        jdf = self._gd1._jgd.flatMapCoGroupsInPandas(jgds, udf_column_expr)
```

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala [hunk 6]
```diff
@@ -582,48 +582,59 @@ class RelationalGroupedDataset protected[sql](
...
       expr: PythonUDF): DataFrame = {
     require(expr.evalType == PythonEvalType.SQL_COGROUPED_MAP_PANDAS_UDF,
       "Must pass a cogrouped map udf")
-    require(this.groupingExprs.length == r.groupingExprs.length,
-      "Cogroup keys must have same size: " +
-        s"${this.groupingExprs.length} != ${r.groupingExprs.length}")
+    val groupingExprLengthEquals = rs.map(_.groupingExprs.length).
+      forall(_ == this.groupingExprs.length)
+
+    require(groupingExprLengthEquals, s"Cogroup keys must have same size.")
```
Comment (EnricoMi) [inline sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala:602]: Could we keep the key sizes in the error message?

Something like `(this +: rs).map(_.groupingExprs.length).mkString(", ")`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala [hunk 7]
```diff
@@ -227,7 +227,7 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
         newVersion.copyTagsFrom(oldVersion)
         Seq((oldVersion, newVersion))
 
-      case oldVersion @ FlatMapCoGroupsInPandas(_, _, _, output, _, _)
+      case oldVersion@FlatMapCoGroupsInPandas(_, _, output, _)
```
Comment (EnricoMi) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:230]: convention is spaces around `@`:
```suggestion
      case oldVersion @ FlatMapCoGroupsInPandas(_, _, output, _)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala [hunk 8]
```diff
@@ -1365,6 +1365,32 @@ trait QuaternaryLike[T <: TreeNode[T]] { self: TreeNode[T] =>
   protected def withNewChildrenInternal(newFirst: T, newSecond: T, newThird: T, newFourth: T): T
 }
 
+trait NaryLike[T <: TreeNode[T]] { self: TreeNode[T] =>
```
Comment (EnricoMi) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:1368]: what is the benefit of defining `NaryLike` over `TreeNode[T]`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/CoGroupedArrowPythonRunner.scala [hunk 9]
```diff
@@ -34,37 +34,32 @@ import org.apache.spark.sql.util.ArrowUtils
...
-  extends BasePythonRunner[
-    (Iterator[InternalRow], Iterator[InternalRow]), ColumnarBatch](funcs, evalType, argOffsets)
-  with BasicPythonArrowOutput {
+                                       funcs: Seq[ChainedPythonFunctions],
+                                       evalType: Int,
+                                       argOffsets: Array[Array[Int]],
+                                       schemas: List[StructType],
+                                       timeZoneId: String,
+                                       conf: Map[String, String],
+                                       val pythonMetrics: Map[String, SQLMetric])
```
Comment (EnricoMi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/CoGroupedArrowPythonRunner.scala:48]:
```suggestion
    funcs: Seq[ChainedPythonFunctions],
    evalType: Int,
    argOffsets: Array[Array[Int]],
    schemas: List[StructType],
    timeZoneId: String,
    conf: Map[String, String],
    val pythonMetrics: Map[String, SQLMetric])
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala [hunk 10]
```diff
@@ -304,7 +304,7 @@ trait FlatMapGroupsWithStateExecBase
           // We apply the values for the key after applying the initial state.
           callFunctionAndUpdateState(
             stateManager.getState(store, keyUnsafeRow),
-              valueRowIter,
+            coGroupedIterators.head,
```
Comment (EnricoMi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala:307]:
```suggestion
              coGroupedIterators.head,
```

**Review by EnricoMi — COMMENTED** [review]
first glance comments, let's wait for committer's thoughts

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala (hunk 6 above)
Comment (santosh-d3vpl3x) [reply sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala:602]: Good suggestion, I will add them.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala (hunk 8 above)
Comment (santosh-d3vpl3x) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:1368]: Consistency with other `QuaternaryLike`, etc. which has been extended with expressions, plans, etc.

**Comment by HyukjinKwon** [comment]
qq,

> this is a breaking change to this experimental API.

What's breaking? would be good to keep the PR desc template (https://github.com/apache/spark/blob/master/.github/PULL_REQUEST_TEMPLATE)

**Comment by santosh-d3vpl3x** [comment]
> qq,
> 
> > this is a breaking change to this experimental API.
> 
> What's breaking? would be good to keep the PR desc template (https://github.com/apache/spark/blob/master/.github/PULL_REQUEST_TEMPLATE)

@HyukjinKwon I have changed the description to include the breaking change and some more details.

**Comment by santosh-d3vpl3x** [comment]
@HyukjinKwon may I request for a review on this PR?

**Comment by EnricoMi** [comment]
CC @cloud-fan @sunchao @zhengruifeng

**Comment by EnricoMi** [comment]
@HyukjinKwon @cloud-fan @sunchao @zhengruifeng this is so sad.

Can we get some feedback if this contribution is welcome or if any more effort is just wasted.

**Pull request closed without merging**